File: android/app/src/main/java/com/example/workflowupdater/ui/main/MainScreen.kt
package com.example.workflowupdater.ui.main

import android.widget.Toast
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CloudOff
import androidx.compose.material.icons.filled.FilterList
import androidx.compose.material.icons.filled.PictureAsPdf
import androidx.compose.material.icons.filled.Search
import androidx.compose.material.icons.filled.SwitchAccount
import androidx.compose.material3.Badge
import androidx.compose.material3.BadgedBox
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ExtendedFloatingActionButton
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TextField
import androidx.compose.material3.TextFieldDefaults
import androidx.compose.material3.TopAppBar
import androidx.compose.material3.TopAppBarDefaults
import androidx.compose.material3.pulltorefresh.PullToRefreshBox
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.example.workflowupdater.data.SheetConfig
import com.example.workflowupdater.pdf.PdfExporter
import com.example.workflowupdater.pdf.rememberPdfExporter

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun MainScreen(viewModel: WorksViewModel, onWorkClick: (Int) -> Unit, modifier: Modifier = Modifier) {
  val context = LocalContext.current
  val state by viewModel.uiState.collectAsStateWithLifecycle()

  val pdfExporter = rememberPdfExporter()

  val filterSheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)
  val profileSheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)
  var showFilterSheet by remember { mutableStateOf(false) }
  var showProfileSheet by remember { mutableStateOf(false) }
  var showExportNameDialog by remember { mutableStateOf(false) }

  Scaffold(
    modifier = modifier.fillMaxSize(),
    topBar = {
      TopAppBar(
        colors =
          TopAppBarDefaults.topAppBarColors(
            containerColor = MaterialTheme.colorScheme.surface,
            titleContentColor = MaterialTheme.colorScheme.onSurface,
          ),
        title = {
          Column {
            Text(text = "RDO KKD Works", style = MaterialTheme.typography.titleLarge, maxLines = 1)
            Text(
              text =
                if (SheetConfig.isAllProfile(state.activeProfile)) {
                  "All engineers \u2022 ${state.filteredWorks.size} shown"
                } else {
                  "Engineer ${state.activeProfile.id} \u2022 ${state.filteredWorks.size} shown"
                },
              style = MaterialTheme.typography.labelMedium,
              color = MaterialTheme.colorScheme.onSurfaceVariant,
              maxLines = 1,
            )
          }
        },
        actions = {
          IconButton(onClick = { showProfileSheet = true }) {
            Icon(Icons.Filled.SwitchAccount, contentDescription = "Switch engineer profile")
          }
          BadgedBox(
            badge = {
              val activeCount = countActiveDropdownFilters(state.filters)
              if (activeCount > 0) Badge { Text(activeCount.toString()) }
            }
          ) {
            IconButton(onClick = { showFilterSheet = true }) {
              Icon(Icons.Filled.FilterList, contentDescription = "Filter works")
            }
          }
        },
      )
    },
    floatingActionButton = {
      ExtendedFloatingActionButton(
        onClick = {
          if (state.isExporting) return@ExtendedFloatingActionButton
          if (state.filteredWorks.isEmpty()) {
            Toast.makeText(context, "No works to export", Toast.LENGTH_SHORT).show()
            return@ExtendedFloatingActionButton
          }
          showExportNameDialog = true
        },
        icon = {
          if (state.isExporting) {
            CircularProgressIndicator(modifier = Modifier.size(20.dp), strokeWidth = 2.dp, color = MaterialTheme.colorScheme.onPrimary)
          } else {
            Icon(Icons.Filled.PictureAsPdf, contentDescription = null)
          }
        },
        text = { Text(if (state.isExporting) "Preparing\u2026" else "Export PDF") },
      )
    },
  ) { innerPadding ->
    Column(modifier = Modifier.padding(innerPadding).fillMaxSize()) {
      SearchField(query = state.searchQuery, onQueryChange = viewModel::onSearchQueryChange)

      Spacer(Modifier.height(8.dp))
      StatChipsRow(
        statusCounts = state.statusCounts,
        selectedCode = state.filters.statusCode,
        onChipClick = viewModel::onStatusChipSelected,
        hasAnyFilter = state.hasAnyFilter,
        onClearAllFilters = viewModel::clearAllFilters,
      )
      if (state.hasAnyFilter) {
        Spacer(Modifier.height(8.dp))
        FilterResultChip(filteredCount = state.filteredWorks.size, totalCount = state.allWorks.size)
      }
      Spacer(Modifier.height(8.dp))

      if (state.isOffline) {
        OfflineBanner(message = state.errorMessage)
      }

      PullToRefreshBox(
        isRefreshing = state.isRefreshing,
        onRefresh = viewModel::refresh,
        modifier = Modifier.fillMaxSize(),
      ) {
        when {
          state.isLoading -> LoadingState()
          state.filteredWorks.isEmpty() -> EmptyState(hasFilters = state.hasAnyFilter, onClear = viewModel::clearAllFilters)
          else ->
            LazyColumn(
              modifier = Modifier.fillMaxSize(),
              contentPadding = PaddingValues(start = 16.dp, end = 16.dp, top = 8.dp, bottom = 96.dp),
              verticalArrangement = Arrangement.spacedBy(12.dp),
            ) {
              items(items = state.filteredWorks, key = { it.rowNum }) { work ->
                WorkCard(work = work, onClick = { onWorkClick(work.rowNum) })
              }
            }
        }
      }
    }
  }

  if (showFilterSheet) {
    FilterSheet(
      sheetState = filterSheetState,
      districtOptions = state.districtOptions,
      lacOptions = state.lacOptions,
      asStatusOptions = state.asStatusOptions,
      arStatusOptions = state.arStatusOptions,
      srStatusOptions = state.srStatusOptions,
      filters = state.filters,
      onApply = viewModel::applyFilters,
      onDismiss = { showFilterSheet = false },
    )
  }

  if (showProfileSheet) {
    ProfileSheet(
      sheetState = profileSheetState,
      activeProfile = state.activeProfile,
      defaultProfileId = state.defaultProfileId,
      onSelect = viewModel::selectProfile,
      onSetDefault = { profile ->
        viewModel.setDefaultProfile(profile)
        Toast.makeText(context, "${profileDisplayName(profile)} set as default on app launch", Toast.LENGTH_SHORT).show()
      },
      onDismiss = { showProfileSheet = false },
    )
  }

  if (showExportNameDialog) {
    ExportPdfNameDialog(
      designation = state.activeProfile.id,
      onDismiss = { showExportNameDialog = false },
      onConfirm = { engineerName ->
        showExportNameDialog = false
        exportPdf(
          context = context,
          state = state,
          viewModel = viewModel,
          pdfExporter = pdfExporter,
          engineerName = engineerName,
        )
      },
    )
  }
}

private fun countActiveDropdownFilters(filters: WorkFilters): Int =
  listOf(filters.district, filters.lac, filters.asStatus, filters.arStatus, filters.srStatus).count { it != null }

private fun exportPdf(
  context: android.content.Context,
  state: WorksUiState,
  viewModel: WorksViewModel,
  pdfExporter: PdfExporter,
  engineerName: String,
) {
  viewModel.setExporting(true)
  val html = viewModel.buildReportHtml(engineerName)
  pdfExporter.printReport(
    html = html,
    jobName = PdfExporter.jobName(state.activeProfile.id, engineerName),
    onStarted = { viewModel.setExporting(false) },
    onError = { message ->
      viewModel.setExporting(false)
      Toast.makeText(context, "Could not create PDF: $message", Toast.LENGTH_LONG).show()
    },
  )
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
private fun SearchField(query: String, onQueryChange: (String) -> Unit) {
  TextField(
    value = query,
    onValueChange = onQueryChange,
    modifier = Modifier.fillMaxWidth().padding(horizontal = 16.dp, vertical = 4.dp),
    placeholder = { Text("Search work, LAC or remarks\u2026") },
    leadingIcon = { Icon(Icons.Filled.Search, contentDescription = null) },
    singleLine = true,
    shape = MaterialTheme.shapes.large,
    colors =
      TextFieldDefaults.colors(
        focusedContainerColor = MaterialTheme.colorScheme.surfaceContainer,
        unfocusedContainerColor = MaterialTheme.colorScheme.surfaceContainer,
        focusedIndicatorColor = androidx.compose.ui.graphics.Color.Transparent,
        unfocusedIndicatorColor = androidx.compose.ui.graphics.Color.Transparent,
        disabledIndicatorColor = androidx.compose.ui.graphics.Color.Transparent,
      ),
  )
}

@Composable
private fun OfflineBanner(message: String?) {
  Row(
    modifier =
      Modifier.fillMaxWidth()
        .padding(horizontal = 16.dp, vertical = 4.dp)
        .background(MaterialTheme.colorScheme.surfaceContainerHigh, MaterialTheme.shapes.small)
        .padding(12.dp),
    verticalAlignment = Alignment.CenterVertically,
  ) {
    Icon(Icons.Filled.CloudOff, contentDescription = null, tint = MaterialTheme.colorScheme.onSurfaceVariant, modifier = Modifier.size(18.dp))
    Spacer(Modifier.width(8.dp))
    Text(
      text = message ?: "Showing saved data \u2014 pull down to retry the live sheet",
      style = MaterialTheme.typography.bodySmall,
      color = MaterialTheme.colorScheme.onSurfaceVariant,
    )
  }
}

@Composable
private fun LoadingState() {
  Box(modifier = Modifier.fillMaxSize(), contentAlignment = Alignment.Center) {
    CircularProgressIndicator(color = MaterialTheme.colorScheme.primary)
  }
}

@Composable
private fun EmptyState(hasFilters: Boolean, onClear: () -> Unit) {
  Box(modifier = Modifier.fillMaxSize().padding(32.dp), contentAlignment = Alignment.Center) {
    Column(horizontalAlignment = Alignment.CenterHorizontally) {
      Text(
        text = if (hasFilters) "No works match your filters" else "No works found for this engineer",
        style = MaterialTheme.typography.titleMedium,
        color = MaterialTheme.colorScheme.onSurface,
        textAlign = TextAlign.Center,
      )
      if (hasFilters) {
        Spacer(Modifier.height(8.dp))
        Text(
          text = "Tap to clear all filters",
          style = MaterialTheme.typography.labelLarge,
          color = MaterialTheme.colorScheme.primary,
          modifier = Modifier.clickable(onClick = onClear),
        )
      }
    }
  }
}

File: docs/works/js/ui/mainScreen.js
/**
 * The works list: top bar, search, KPI status chips, filter/profile sheets, offline banner,
 * pull-to-refresh and the Export PDF action.
 * Ported from android/.../ui/main/MainScreen.kt.
 */

import { isAllProfile } from '../config.js';
import { countActiveDropdownFilters, hasAnyFilter } from '../state.js';
import { clear, el, iconButton } from './dom.js';
import { Icons } from './icons.js';
import { createFilterResultChip, createStatChipsRow } from './chips.js';
import { showFilterSheet } from './filterSheet.js';
import { profileDisplayName, showProfileSheet } from './profileSheet.js';
import { showExportPdfNameDialog } from './exportDialog.js';
import { jobNameFor, printReport } from './pdfExport.js';
import { showToast } from './toast.js';
import { workCard } from './workCard.js';
import { attachPullToRefresh } from './pullToRefresh.js';

export function createMainScreen({ viewModel, onWorkClick }) {
  let state = viewModel.getState();

  const subtitle = el('p', { className: 'app-bar-subtitle' });
  const filterBadge = el('span', { className: 'filter-badge', hidden: true });

  const profileButton = iconButton(Icons.switchAccount(), {
    label: 'Switch engineer profile',
    onClick: () => showProfileSheet({
      state,
      onSelect: (profile) => viewModel.selectProfile(profile),
      onSetDefault: (profile) => {
        viewModel.setDefaultProfile(profile);
        showToast(`${profileDisplayName(profile)} set as default on app launch`);
      },
    }),
  });

  const refreshButton = iconButton(Icons.refresh(), {
    label: 'Refresh from the live sheet',
    className: 'refresh-btn',
    onClick: () => viewModel.refresh(),
  });

  const filterButton = iconButton(Icons.filterList(), {
    label: 'Filter works',
    onClick: () => showFilterSheet({ state, onApply: (filters) => viewModel.applyFilters(filters) }),
  });
  filterButton.append(filterBadge);

  const appBar = el('header', { className: 'app-bar' }, [
    el('div', { className: 'app-bar-titles' }, [
      el('h1', { className: 'app-bar-title', text: 'RDO KKD Works' }),
      subtitle,
    ]),
    el('div', { className: 'app-bar-actions' }, [profileButton, refreshButton, filterButton]),
  ]);

  const searchInput = el('input', {
    className: 'search-input',
    attrs: {
      type: 'search',
      placeholder: 'Search work, LAC or remarks…',
      'aria-label': 'Search work, LAC or remarks',
      enterkeyhint: 'search',
    },
    on: { input: (event) => viewModel.onSearchQueryChange(event.target.value) },
  });

  const searchField = el('div', { className: 'search-field' }, [
    el('span', { className: 'search-icon', html: Icons.search() }),
    searchInput,
  ]);

  const chipsRow = createStatChipsRow({
    onChipClick: (code) => viewModel.onStatusChipSelected(code),
    onStatusOrderChange: (order) => viewModel.onStatusChipOrderChange(order),
    onClearAllFilters: () => viewModel.clearAllFilters(),
  });

  const filterResultChip = createFilterResultChip();
  const offlineBanner = el('div', { className: 'offline-banner', hidden: true });
  const list = el('div', { className: 'works-list' });
  const refreshIndicator = el('div', { className: 'refresh-indicator', hidden: true }, [el('span', { className: 'spinner' })]);
  const scrollArea = el('main', { className: 'scroll-area' }, [refreshIndicator, list]);

  const exportFab = el('button', {
    className: 'fab',
    attrs: { type: 'button' },
    on: { click: onExportClick },
  }, [el('span', { className: 'fab-icon', html: Icons.pictureAsPdf() }), el('span', { className: 'fab-label', text: 'Export PDF' })]);

  const root = el('div', { className: 'screen main-screen' }, [
    appBar,
    el('div', { className: 'main-controls' }, [searchField, chipsRow.root, filterResultChip.root, offlineBanner]),
    scrollArea,
    exportFab,
  ]);

  attachPullToRefresh(scrollArea, () => viewModel.refresh());

  function onExportClick() {
    if (state.isExporting) return;
    if (state.filteredWorks.length === 0) {
      showToast('No works to export');
      return;
    }
    showExportPdfNameDialog({
      designation: state.activeProfile.id,
      onConfirm: (engineerName) => exportPdf(engineerName),
    });
  }

  function exportPdf(engineerName) {
    viewModel.setExporting(true);
    const html = viewModel.buildReportHtml(engineerName);
    printReport({
      html,
      jobName: jobNameFor(state.activeProfile.id, engineerName),
      onStarted: () => viewModel.setExporting(false),
      onError: (message) => {
        viewModel.setExporting(false);
        showToast(`Could not create PDF: ${message}`, { long: true });
      },
    });
  }

  function renderOfflineBanner() {
    if (!state.isOffline) {
      offlineBanner.hidden = true;
      return;
    }
    const syncedLabel = state.lastSyncedAtMillis
      ? `Last updated ${new Date(state.lastSyncedAtMillis).toLocaleString(undefined, {
          day: '2-digit',
          month: 'short',
          year: 'numeric',
          hour: 'numeric',
          minute: '2-digit',
        })}`
      : null;

    let body;
    if (state.errorMessage && syncedLabel) body = `${state.errorMessage} · ${syncedLabel}`;
    else if (state.errorMessage) body = state.errorMessage;
    else if (syncedLabel) body = `Showing saved data — ${syncedLabel}. Pull down to retry.`;
    else body = 'Showing saved data — pull down to retry the live sheet';

    clear(offlineBanner);
    offlineBanner.append(el('span', { className: 'meta-icon', html: Icons.cloudOff() }), el('span', { text: body }));
    offlineBanner.hidden = false;
  }

  function renderList() {
    clear(list);

    if (state.isLoading) {
      list.append(el('div', { className: 'state-block' }, [el('span', { className: 'spinner large' })]));
      return;
    }

    if (state.filteredWorks.length === 0) {
      const anyFilter = hasAnyFilter(state);
      const block = el('div', { className: 'state-block' }, [
        el('p', {
          className: 'empty-title',
          text: anyFilter ? 'No works match your filters' : 'No works found for this engineer',
        }),
      ]);
      if (anyFilter) {
        block.append(
          el('button', {
            className: 'link-btn',
            text: 'Tap to clear all filters',
            attrs: { type: 'button' },
            on: { click: () => viewModel.clearAllFilters() },
          }),
        );
      }
      list.append(block);
      return;
    }

    const fragment = document.createDocumentFragment();
    for (const work of state.filteredWorks) fragment.append(workCard(work, onWorkClick));
    list.append(fragment);
  }

  function render(next) {
    state = next;

    subtitle.textContent = isAllProfile(state.activeProfile)
      ? `All engineers • ${state.filteredWorks.length} shown`
      : `Engineer ${state.activeProfile.id} • ${state.filteredWorks.length} shown`;

    if (searchInput.value !== state.searchQuery) searchInput.value = state.searchQuery;

    const activeCount = countActiveDropdownFilters(state.filters);
    filterBadge.textContent = String(activeCount);
    filterBadge.hidden = activeCount === 0;

    chipsRow.render({ ...state, hasAnyFilter: hasAnyFilter(state) });

    if (hasAnyFilter(state)) {
      filterResultChip.render(state.filteredWorks.length, state.allWorks.length);
      filterResultChip.root.hidden = false;
    } else {
      filterResultChip.root.hidden = true;
    }

    renderOfflineBanner();
    refreshIndicator.hidden = !state.isRefreshing;
    refreshButton.classList.toggle('spinning', state.isRefreshing);

    exportFab.classList.toggle('busy', state.isExporting);
    exportFab.querySelector('.fab-label').textContent = state.isExporting ? 'Preparing…' : 'Export PDF';
    exportFab.querySelector('.fab-icon').innerHTML = state.isExporting ? '<span class="spinner"></span>' : Icons.pictureAsPdf();

    renderList();
  }

  return { root, render };
}
